### PR TITLE
feat: proxy-mirror plugin resolve host

### LINKED ISSUE #7796: feat: As a user, I want to let proxy-mirror plugin resolve the host, so that the DNS resolver in APISIX can be used.
### Description

The proxy-mirror doesn't support brighter DNS resolution, as its feature is currently based on Nginx's directive, which uses Nginx's own DNS resolution.

As Nginx's own DNS resolution doesn't support modern DNS features, this brings some inconvenience, especially under k8s.

To solve the problem, we need to resolve the host ourselves, and pass the result to the Nginx variable `upstream_mirror_uri`.
https://github.com/apache/apisix/blob/3c0757477a8c4e24910bb998d06e1d13ec283738/apisix/plugins/proxy-mirror.lua#L63-L65

We can use the method below to resolve the host:
https://github.com/apache/apisix/blob/3c0757477a8c4e24910bb998d06e1d13ec283738/apisix/core/resolver.lua#L44

### PR DESCRIPTION
**Pull request by Strangevy**
### Description

Fixes #7796 

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/proxy-mirror.lua [hunk 1]
```diff
@@ -59,9 +61,22 @@ function _M.check_schema(conf)
 end
 
 
+local function resolver_host(prop_host)
+    local url_decoded = url.parse(prop_host)
+    if not core.utils.parse_ipv4(url_decoded.host) and not core.utils.parse_ipv6(url_decoded.host) then
+        local ip = core.resolver.parse_domain(url_decoded.host)
```
Comment (spacewander) [inline apisix/plugins/proxy-mirror.lua:67]: The parse_domain may return `err`, let's check for it

File: t/plugin/proxy-mirror.t [hunk 2]
```diff
@@ -696,3 +696,49 @@ passed
...
+       }
+--- response_body
+passed
+
+
+
+=== TEST 24: hit route resolver domain
+--- request
+GET /hello
+--- error_code: 200
```
Comment (spacewander) [inline t/plugin/proxy-mirror.t:740]: We don't need to check `error_code: 200` as it is checked by default

File: t/plugin/proxy-mirror.t [hunk 3]
```diff
@@ -696,3 +696,49 @@ passed
...
+--- response_body
+hello world
+--- error_log
+http://httpbin.org is resolved to:
```
Comment (spacewander) [inline t/plugin/proxy-mirror.t:744]: Let's use a regex to validate it is indeed resolved to an IP

File: apisix/plugins/proxy-mirror.lua [hunk 4]
```diff
@@ -59,9 +61,30 @@ function _M.check_schema(conf)
 end
 
 
+local function resolver_host(prop_host)
+    local url_decoded = url.parse(prop_host)
+    local decoded_host = url_decoded.host
+    if not core.utils.parse_ipv4(decoded_host) and not core.utils.parse_ipv6(decoded_host) then
+        local ip, err = core.resolver.parse_domain(decoded_host)
+
+        if err then
```
Comment (spacewander) [inline apisix/plugins/proxy-mirror.lua:70]: For code style, I would prefer to use `if not ip`

File: apisix/plugins/proxy-mirror.lua [hunk 5]
```diff
@@ -59,9 +61,30 @@ function _M.check_schema(conf)
...
+            core.log.error("dns resolver resolves domain: ", decoded_host, " error: ", err,
+             " will continue to use the host: ", decoded_host)
```
Comment (spacewander) [inline apisix/plugins/proxy-mirror.lua:72]: The indentation can be better

**Review by tzssangglass — APPROVED** [review]

**Comment by spacewander** [comment]
@Strangevy 
Please make the CI pass, thanks!

**Comment by Strangevy** [comment]
> @Strangevy Please make the CI pass, thanks!

Done

**Review by spacewander — APPROVED** [review]

**Review by starsz — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**